include/fi_indexer.h

```diff
@@ -60,19 +60,26 @@ struct indexer
 {
 	union ofi_idx_entry *array[OFI_IDX_ARRAY_SIZE];
 	int		 free_list;
-	int		 size;
+	int		 size;	/* Array size (used): [0, OFI_IDX_ARRAY_SIZE) */
 };
 
 #define ofi_idx_array_index(index) (index >> OFI_IDX_ENTRY_BITS)
 #define ofi_idx_entry_index(index) (index & (OFI_IDX_ENTRY_SIZE - 1))
 
 int ofi_idx_insert(struct indexer *idx, void *item);
 void *ofi_idx_remove(struct indexer *idx, int index);
-void ofi_idx_replace(struct indexer *idx, int index, void *item);
+int ofi_idx_replace(struct indexer *idx, int index, void *item);
 void ofi_idx_reset(struct indexer *idx);
 
+static inline int ofi_idx_is_valid(struct indexer *idx, int index)
+{
+	return (index > 0) && (index < idx->size * OFI_IDX_ENTRY_SIZE);
+}
+
 static inline void *ofi_idx_at(struct indexer *idx, int index)
 {
+	if (!ofi_idx_is_valid(idx, index))
+		return NULL;
 	return (idx->array[ofi_idx_array_index(index)] + ofi_idx_entry_index(index))->item;
 }
```

Comment:
Please add a new function for this, ofi_idx_lookup().  This will mirror the idm code, where 'lookup' does validation, but 'at' just returns the entry, so we can avoid the validation check when it's unnecessary.